### ISSUE TITLE
Add Invitation for registration

### ISSUE THREAD

**Issue by xundeenergie** [LemmyNet/lemmy#1777]
For front end issues, use [lemmy-ui](https://github.com/LemmyNet/lemmy-ui)

### Is your proposal related to a problem?

I don't know, if it's a frontend- or a backendissue... Registration in lemmy is very "basic". Open registration invites spammers to spam. Ok. Now there are captures... 

But i can not have closed registration and create manually a user.

### Describe the solution you'd like

Add "invitations" to login-process

For my usecase i will prefer "Invites". Create a link which is valid for a due-date and it can be used only $COUNT times. Pleroma has this feature, and it is great. I can create for example one invitation-link with count 4 and or without duedate.
so i can send this link in my family-group (4 members except me), and they can use ist. if they give it to someone i don't like to have on my server, i wil recognize ist.

### Describe alternatives you've considered

Make it possible, to create a user manually as admin, even with closed registration.

**Comment by dessalines** [comment]
You can close account creation, that's in your instance sidebar settings, and it also asks you that when you first set up lemmy. 

Doing a chain invite system is a separate thing though. In the meantime you can just set up accts for people and send them the login.

**Comment by Nothing4You** [comment]
supporting an invite model also has the great benefit of being more privacy preserving by making it more viable for instance admins to not require email addresses from users, which even reddit still supports.

**Comment by sunaurus** [comment]
I would really like to tackle this issue and put together a PR, but I'm quite busy for the next week or so. If nobody else begins work on this then I could probably do so early next month! 

My general approach would be:
* Admins can require invites for new sign-ups, this will add a new "invite code" field to the sign-up form (which can also be pre-filled from an invite link)
* The max amount of invites a user can automatically accumulate would be capped (admins could configure this cap)
* Users would accumulate invites at some reasonable pace every month - maybe (max_amount/2)?

**Comment by lionirdeadman** [comment]
I think it would be good to have an invite system *in addition* to the application system.

That way, people's friends can be more easily invited and we get less backlog for application and for people who don't know friends who use the instance, they can still go through the application system.

**Comment by ilovethensa** [comment]
Any news on this?

**Comment by dessalines** [comment]
No one above is working on it, but anyone is free to do so.

**Comment by Nutomic** [comment]
@kryoseu has started to work on this. It is still a bit unclear how exactly it should be configured. Should this be a new RegistrationMode value, and only if the admin explicitly sets RegistrationMode.InviteOnly can users generate invite links? Or should it be a separate setting, so that users can generate invite links regardless of RegistrationMode? (eg with open and also closed registrations)

**Comment by kryoseu** [comment]
> Or should it be a separate setting, so that users can generate invite links regardless of RegistrationMode? (eg with open and also closed registrations)

I'm inclined to believe there should be an option for admins to completely lock the instance, including the ability to issue new invitations. In which case invitation links should work for all registration modes except Close.

**Comment by MrKaplan-lw** [comment]
imo admins should always be able to create invitations. since admins can change this setting, it doesn't seem like there's any point in restricting admins from using them at any point. what would need to be controlled is whether users can create invites.

admins should also be able to see pending invites and their creators. it is crucial to track who created invites and who someone was invited by to be able to deal with abuse.
users should be able to see all invites they created and be able to revoke any created but not redeemed invites.

nice to have things
- admins should be able to restrict selected users from being able to create invites (denylist)
- admins should be able to restrict who can create invites (allowlist)
- limits for how many times invites can be used, up to unlimited. max being defined on instance level, while allowing individual invites to have lower limits defined at creation time. admins should be able to create invites with any limit, even exceeding the instance limit.
- limits on how many other users a user can invite. this could be a permanent limit or maybe something that refreshes e.g. monthly. bonus points if admins could raise the limit for selected users.
- some sort of activity requirements to be eligible to create invites, such as a certain number of posts or comments
- easy visualization of invite chain for admins, as well as tree view for invitees by a user and their invitees

**Comment by Nutomic** [comment]
> imo admins should always be able to create invitations. since admins can change this setting, it doesn't seem like there's any point in restricting admins from using them at any point. what would need to be controlled is whether users can create invites.

Agree

> admins should also be able to see pending invites and their creators. it is crucial to track who created invites and who someone was invited by to be able to deal with abuse.
users should be able to see all invites they created and be able to revoke any created but not redeemed invites.

True, although these dont have to be part of the initial pull request. There will be enough time to add them later, before the final 1.0 release. Same goes for your nice to have things.

**Comment by kryoseu** [comment]
We could have a setting under "Registration mode" for "Allow invitations", so registration modes stay as is and admins can easily enable/disable invitations. 

For the initial PR:
- Admins can toggle invitations on and off from settings.
- Admins and users can create invitation links.
- Admins and users can list and revoke their active invitations.
- Each invitation link can have max uses (up to unlimited or instance-level limit) and expiration.

> admins should also be able to see pending invites and their creators

I believe the idea is for users to be able to create invite links, in which case there is no concept of a "pending invite". Links can be active, revoked or exhausted. I think the closest would be to make admins able to see all users who registered from invitations and who they were invited by.

**Comment by MrKaplan-lw** [comment]
pending invites = unredeemed invite links.

another nice to have later on could be allowing users to invite others via email, which is something piefed already supports (although i don't know if this is just a link to sign up or linked to a redemption system), but this would need careful consideration to prevent abuse, for rate limits.

**Comment by kryoseu** [comment]
> pending invites = unredeemed invite links.

Sorry, I'm still a bit confused here. I think there's some misunderstanding around this. 

The invite link either gets used or it doesn't. A "pending/unredeemed" state only makes sense when an invite is sent to a specific person (e.g. via email) and tracked until redeemed. The issue here is about creating Mastodon-like invite link system — they're stateless tokens with optional use-count and expiry limits. There's no pending state at the invite level.

**Comment by MrKaplan-lw** [comment]
an invite is unredeemed at the point where it is created, until it gets redeemed by someone signing up (or a counter decremented if it permits reuse, or never if unlimited uses).

**Comment by kryoseu** [comment]
> an invite is unredeemed at the point where it is created, until it gets redeemed by someone signing up (or a counter decremented if it permits reuse, or never if unlimited uses).

I think we're both talking about the same thing, just using different words. What you're calling "pending/unredeemed" I'm calling "Active", since "pending" implies something is waiting on a person, whereas "active" describes the invite link's state.

Appreciate the clarification.

**Comment by db0** [comment]
Shouldn't this be closed once the PR is merged?

**Comment by Nutomic** [comment]
The backend PR is already merged (https://github.com/LemmyNet/lemmy/pull/6450). For frontend there is a separate issue (https://github.com/LemmyNet/lemmy-ui/issues/4128), and PR is work in progress (https://github.com/LemmyNet/lemmy-ui/pull/4298).
